script.lua

```diff
@@ -5610,6 +5610,52 @@ function upCastPosSizRot(pos,size,rot,dist,types)
     end
     return hitObjects
 end
+-----
+local Elements = {}
+Elements.__index = Elements
+function Elements:new(init)
+    local outTable = {0,0,0,0,0,0,0,0}
+    setmetatable(outTable, self)
+    outTable:add(init)
+    return outTable
+end
+function Elements:add(other)
+    if other == nil then
+        return
+    elseif type(other) == "table" then
+        for i = 1, 8 do
+            self[i] = self[i] + other[i]
+        end
+    elseif type(other) == "string" then
+        for i = 1, string.len(other) do
+            self[i] = self[i] + math.floor(string.sub(other, i, i))
+        end
+    end
+end
+function Elements:threshold(threshold)
+    for i = 1, 8 do
+        if threshold[i] > self[i] then
+            return false
+        end
+    end
+    return true
+end
+function Elements:__tostring()
+    return table.concat(self, "")
+end
+
+local function modifyElements(params)
+    for _,object in pairs(getObjectsWithTag("Modify Elements")) do
+        params.elements = object.call("modifyElements", params)
+    end
+    return Elements:new(params.elements)
```

Comment:
do we need to call `Elements:new` again? we called it before to pass in parameter to `modifyElements`. is this just kind of sanity checking the return value of from the object function callss?

Reply:
People writing their own `modifyElements()` functions won't have access to the `Elements` metatable (or more specifically, won't have access to the functions *in* the `Elements` metatable, even if they can access the metatable itself). And I haven't even mentioned the metatable in the documentation. That's no concern if they modify and return `params.elements`, 'cause then the metatable will still be attached. But if they decide to build the table they return from scratch (which the documentation doesn't tell them not to do), then it'll be a plain table without the elements metatable attached. Calling `Elements:new()` on it reattaches the metatable, making it essentially a proper `Elements` "object" again, rather than just a plain table.

Reply:
And it's necessary to ensure that the object has the metatable attached, 'cause `addThresholdDecals()` later uses `elements:threshold()` from the metatable: https://github.com/spirit-island/spirit-island-mod/blob/083daedc00cea36acefcef1889b87ed3b06d3439/script.lua#L5811

Reply:
makes sense, thanks